tls/s2n_x509_validator.c

```diff
@@ -0,0 +1,446 @@
+/*
+ * Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/s2n_config.h"
+#include "utils/s2n_asn1_time.h"
+#include "utils/s2n_safety.h"
+#include "tls/s2n_connection.h"
+
+#include "openssl/err.h"
+#include "openssl/asn1.h"
+
+#if !defined(OPENSSL_IS_BORINGSSL)
+#include "openssl/ocsp.h"
+#endif
+
+/* one day, boringssl, may add ocsp stapling support. Let's future proof this a bit by grabbing a definition
+ * that would have to be there when they add support */
+#if defined(OPENSSL_IS_BORINGSSL) && !defined(OCSP_RESPONSE_STATUS_SUCCESSFUL)
+#define S2N_OCSP_STAPLING_SUPPORTED 0
+#else
+#define S2N_OCSP_STAPLING_SUPPORTED 1
+#endif /* defined(OPENSSL_IS_BORINGSSL) && !defined(OCSP_RESPONSE_STATUS_SUCCESSFUL) */
+
+/* our friends at openssl love to make backwards incompatible changes */
+#if !defined(LIBRESSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER >= 0x10100000L
+#define OCSP_GET_CERTS(a) OCSP_resp_get0_certs(a)
+#else
+#define OCSP_GET_CERTS(a) a->certs
+#endif
+
+uint8_t s2n_x509_ocsp_stapling_supported(void) {
+    return S2N_OCSP_STAPLING_SUPPORTED;
+}
+
+void s2n_x509_trust_store_init(struct s2n_x509_trust_store *store) {
+    store->trust_store = NULL;
+}
+
+uint8_t s2n_x509_trust_store_has_certs(struct s2n_x509_trust_store *store) {
+    return store->trust_store ? (uint8_t) 1 : (uint8_t) 0;
```

Comment:
Is this a complete implemenation? Is there a code path that initializes `store->trust_store` with `X509_STORE_new()` but doesn't add any certs?

Reply:
Yes, but in that case, the loading of the trust store failed for some reason and we want the validator to run as normal and fail.